gufe/protocols/protocolunit.py

```diff
@@ -0,0 +1,85 @@
+import abc
+from typing import Iterable, List
+
+from .results import ProtocolUnitResult, ProtocolResult
+
+
+class ProtocolUnit(abc.ABC):
+    """A unit of work computable by
+
+    """
+
+    def __init__(
+            self,
+            settings: "ProtocolSettings",
+            *dependencies: Iterable["ProtocolUnit"],
+        ):
+
+        self._settings = settings
+        self._dependencies = dependencies
+        self._status = "WAITING"
+
+
+    def execute(self, block=True):
+        """Given `ProtocolUnitResult`s from dependencies, execute this `ProtocolUnit`.
+
+        Parameters
+        ----------
+        block : bool
+            If `True`, block until execution completes; otherwise run in its own thread.
+
+        """
+        if block:
+            dep_results = [dep.estimate() for dep in self._dependencies]
+
+            self._status = "RUNNING"
+            self._execute(dep_results)
+            self._status = "COMPLETE"
```

Comment:
At the risk of one class too many, I think a status class could be handy, could just be a named tuple with fields like status enum(like), datetime started, datetime finished, host id, which then gets stored alongside the produced artifacts(?) as provenance.

Reply:
Actually I think I hate this idea after all.